dali/kernels/imgproc/convolution/filter_gpu.cuh

```diff
@@ -0,0 +1,1094 @@
+    for (int y = block_setup.flat_idx(); y < sample_desc.shape.in_workspace_extents.y;
+         y += block_setup.flat_size()) {
+      precomputed_idx[y] = in_loader.border_remap(anchored_start.y + y, sample_desc.shape, 1);
+    }
+  }
+
+  DALI_DEVICE DALI_FORCEINLINE void precompute_indices(const In* __restrict__ in,
+                                                       const ivec3& anchored_start) const {
+    int* ys = precomputed_idx;
+    for (int y = block_setup.flat_idx(); y < sample_desc.shape.in_workspace_extents.y;
+         y += block_setup.flat_size()) {
+      ys[y] = in_loader.border_remap(anchored_start.y + y, sample_desc.shape, 1);
+    }
+    int* zs = precomputed_idx + sample_desc.shape.in_workspace_extents.y;
+    for (int z = block_setup.flat_idx(); z < sample_desc.shape.in_workspace_extents.z;
+         z += block_setup.flat_size()) {
+      zs[z] = in_loader.border_remap(anchored_start.z + z, sample_desc.shape, 2);
+    }
+  }
+
+  DALI_DEVICE DALI_FORCEINLINE void load_input_to_shm(const In* __restrict__ in,
+                                                      const ivec2& anchored_start) const {
+    const auto& block_dim = block_setup.block_dim();
+    const auto& thread_idx = block_setup.thread_idx();
+    for (int x = thread_idx.x; x < sample_desc.shape.in_workspace_extents.x; x += block_dim.x) {
+      int global_x = in_loader.border_remap_innermost(anchored_start.x + x, sample_desc.shape);
+      for (int y = thread_idx.y; y < sample_desc.shape.in_workspace_extents.y; y += block_dim.y) {
+        int global_y = precomputed_idx[y];
+        in_workspace[dot(ivec2{x, y}, sample_desc.shape.in_workspace_strides)] =
+            in_loader.load(in, ivec2{global_x, global_y}, sample_desc.shape);
+      }
+    }
+  }
+
+  DALI_DEVICE DALI_FORCEINLINE void load_input_to_shm(const In* __restrict__ in,
+                                                      const ivec3& anchored_start) const {
+    const int* const __restrict__ ys = precomputed_idx;
+    const int* const __restrict__ zs = precomputed_idx + sample_desc.shape.in_workspace_extents.y;
+    const auto& block_dim = block_setup.block_dim();
+    const auto& thread_idx = block_setup.thread_idx();
+    for (int x = thread_idx.x; x < sample_desc.shape.in_workspace_extents.x; x += block_dim.x) {
+      int global_x = in_loader.border_remap_innermost(anchored_start.x + x, sample_desc.shape);
+      for (int y = thread_idx.y; y < sample_desc.shape.in_workspace_extents.y; y += block_dim.y) {
+        int global_y = ys[y];
+        for (int z = thread_idx.z; z < sample_desc.shape.in_workspace_extents.z; z += block_dim.z) {
+          int global_z = zs[z];
+          in_workspace[dot(ivec3{x, y, z}, sample_desc.shape.in_workspace_strides)] =
+              in_loader.load(in, ivec3{global_x, global_y, global_z}, sample_desc.shape);
+        }
+      }
+    }
+  }
+
+  const SampleDescT& sample_desc;
+  const Inloader& in_loader;
+  const BlockSetupT& block_setup;
```

Comment:
With all those references it's likely to end up reaching to memory more than you'd wish.